File: projetos-fenix-para-estudo/t2ti-erp-fenix-cadastros-completo/fontes/backend/delphi/dmvc/Controller/Financeiro/FinConfiguracaoBoletoController.pas
{*******************************************************************************
Title: T2Ti ERP Fenix                                                                
Description: Controller relacionado à tabela [FIN_CONFIGURACAO_BOLETO] 
                                                                                
The MIT License                                                                 
                                                                                
Copyright: Copyright (C) 2020 T2Ti.COM                                          
                                                                                
Permission is hereby granted, free of charge, to any person                     
obtaining a copy of this software and associated documentation                  
files (the "Software"), to deal in the Software without                         
restriction, including without limitation the rights to use,                    
copy, modify, merge, publish, distribute, sublicense, and/or sell               
copies of the Software, and to permit persons to whom the                       
Software is furnished to do so, subject to the following                        
conditions:                                                                     
                                                                                
The above copyright notice and this permission notice shall be                  
included in all copies or substantial portions of the Software.                 
                                                                                
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,                 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES                 
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                        
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT                     
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,                    
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING                    
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR                   
OTHER DEALINGS IN THE SOFTWARE.                                                 
                                                                                
       The author may be contacted at:                                          
           t2ti.com@gmail.com                                                   
                                                                                
@author Albert Eije (alberteije@gmail.com)                    
@version 1.0.0
*******************************************************************************}
unit FinConfiguracaoBoletoController;

interface

uses mvcframework, mvcframework.Commons,
  System.SysUtils,
  MVCFramework.SystemJSONUtils;

type

  [MVCDoc('CRUD FinConfiguracaoBoleto')]
  [MVCPath('/fin-configuracao-boleto')]
  TFinConfiguracaoBoletoController = class(TMVCController)
  public
    [MVCDoc('Retorna uma lista de objetos')]
    [MVCPath('/($filtro)')]
    [MVCHTTPMethod([httpGET])]
    procedure ConsultarLista(Context: TWebContext);

    [MVCDoc('Retorna um objeto com base no ID')]
    [MVCPath('/($id)')]
    [MVCHTTPMethod([httpGET])]
    procedure ConsultarObjeto(id: Integer);

    [MVCDoc('Insere um novo objeto')]
    [MVCPath]
    [MVCHTTPMethod([httpPOST])]
    procedure Inserir(Context: TWebContext);

    [MVCDoc('Altera um objeto com base no ID')]
    [MVCPath('/($id)')]
    [MVCHTTPMethod([httpPUT])]
    procedure Alterar(id: Integer);

    [MVCDoc('Exclui um objeto com base no ID')]
    [MVCPath('/($id)')]
    [MVCHTTPMethod([httpDelete])]
    procedure Excluir(id: Integer);
  end;

implementation

{ TFinConfiguracaoBoletoController }

uses FinConfiguracaoBoletoService, FinConfiguracaoBoleto, Commons, Filtro;

procedure TFinConfiguracaoBoletoController.ConsultarLista(Context: TWebContext);
var
  FiltroUrl, FilterWhere: string;
  FiltroObj: TFiltro;
begin
  FiltroUrl := Context.Request.Params['filtro'];
  if FiltroUrl <> '' then
  begin
    ConsultarObjeto(StrToInt(FiltroUrl));
    exit;
  end;

  filterWhere := Context.Request.Params['filter'];
  try
    if FilterWhere = '' then
    begin
      Render<TFinConfiguracaoBoleto>(TFinConfiguracaoBoletoService.ConsultarLista);
    end
    else begin
      // define o objeto filtro
      FiltroObj := TFiltro.Create(FilterWhere);
      Render<TFinConfiguracaoBoleto>(TFinConfiguracaoBoletoService.ConsultarListaFiltroValor(FiltroObj.Campo, FiltroObj.Valor));
    end;
  except
    on E: EServiceException do
    begin
      raise EMVCException.Create
        ('Erro no Servidor [Consultar Lista FinConfiguracaoBoleto] - Exceção: ' + E.Message,
        E.StackTrace, 0, 500);
    end
    else
      raise;
  end;
end;

procedure TFinConfiguracaoBoletoController.ConsultarObjeto(id: Integer);
var
  FinConfiguracaoBoleto: TFinConfiguracaoBoleto;
begin
  try
    FinConfiguracaoBoleto := TFinConfiguracaoBoletoService.ConsultarObjeto(id);

    if Assigned(FinConfiguracaoBoleto) then
      Render(FinConfiguracaoBoleto)
    else
      raise EMVCException.Create
        ('Registro não localizado [Consultar Objeto FinConfiguracaoBoleto]', '', 0, 404);
  except
    on E: EServiceException do
    begin
      raise EMVCException.Create
        ('Erro no Servidor [Consultar Objeto FinConfiguracaoBoleto] - Exceção: ' + E.Message,
        E.StackTrace, 0, 500);
    end
    else
      raise;
  end;
end;

procedure TFinConfiguracaoBoletoController.Inserir(Context: TWebContext);
var
  FinConfiguracaoBoleto: TFinConfiguracaoBoleto;
begin
  try
    FinConfiguracaoBoleto := Context.Request.BodyAs<TFinConfiguracaoBoleto>;
  except
    on E: EServiceException do
    begin
      raise EMVCException.Create('Objeto inválido [Inserir FinConfiguracaoBoleto] - Exceção: ' +
        E.Message, E.StackTrace, 0, 400);
    end
    else
      raise;
  end;

  try
    TFinConfiguracaoBoletoService.Inserir(FinConfiguracaoBoleto);
    Render(TFinConfiguracaoBoletoService.ConsultarObjeto(FinConfiguracaoBoleto.Id));
  finally
  end;
end;

procedure TFinConfiguracaoBoletoController.Alterar(id: Integer);
var
  FinConfiguracaoBoleto, FinConfiguracaoBoletoDB: TFinConfiguracaoBoleto;
begin
  try
    FinConfiguracaoBoleto := Context.Request.BodyAs<TFinConfiguracaoBoleto>;
  except
    on E: EServiceException do
    begin
      raise EMVCException.Create('Objeto inválido [Alterar FinConfiguracaoBoleto] - Exceção: ' +
        E.Message, E.StackTrace, 0, 400);
    end
    else
      raise;
  end;

  if FinConfiguracaoBoleto.Id <> id then
    raise EMVCException.Create('Objeto inválido [Alterar FinConfiguracaoBoleto] - ID do objeto difere do ID da URL.',
      '', 0, 400);

  FinConfiguracaoBoletoDB := TFinConfiguracaoBoletoService.ConsultarObjeto(FinConfiguracaoBoleto.id);

  if not Assigned(FinConfiguracaoBoletoDB) then
    raise EMVCException.Create('Objeto com ID inválido [Alterar FinConfiguracaoBoleto]',
      '', 0, 400);

  try
    if TFinConfiguracaoBoletoService.Alterar(FinConfiguracaoBoleto) > 0 then
      Render(TFinConfiguracaoBoletoService.ConsultarObjeto(FinConfiguracaoBoleto.Id))
    else
      raise EMVCException.Create('Nenhum registro foi alterado [Alterar FinConfiguracaoBoleto]',
        '', 0, 500);
  finally
    FreeAndNil(FinConfiguracaoBoletoDB);
  end;
end;

procedure TFinConfiguracaoBoletoController.Excluir(id: Integer);
var
  FinConfiguracaoBoleto: TFinConfiguracaoBoleto;
begin
  FinConfiguracaoBoleto := TFinConfiguracaoBoletoService.ConsultarObjeto(id);

  if not Assigned(FinConfiguracaoBoleto) then
    raise EMVCException.Create('Objeto com ID inválido [Excluir FinConfiguracaoBoleto]',
      '', 0, 400);

  try
    if TFinConfiguracaoBoletoService.Excluir(FinConfiguracaoBoleto) > 0 then
      Render(200, 'Objeto excluído com sucesso.')
    else
      raise EMVCException.Create('Nenhum registro foi excluído [Excluir FinConfiguracaoBoleto]',
        '', 0, 500);
  finally
    FreeAndNil(FinConfiguracaoBoleto)
  end;
end;

end.

File: projetos-fenix-para-estudo/t2ti-erp-fenix-cadastros-completo/fontes/backend/c-sharp/T2TiERPFenix/T2TiERPFenix/Controllers/Financeiro/FinConfiguracaoBoletoController.cs
/*******************************************************************************
Title: T2Ti ERP Fenix                                                                
Description: Controller relacionado à tabela [FIN_CONFIGURACAO_BOLETO] 
                                                                                
The MIT License                                                                 
                                                                                
Copyright: Copyright (C) 2020 T2Ti.COM                                          
                                                                                
Permission is hereby granted, free of charge, to any person                     
obtaining a copy of this software and associated documentation                  
files (the "Software"), to deal in the Software without                         
restriction, including without limitation the rights to use,                    
copy, modify, merge, publish, distribute, sublicense, and/or sell               
copies of the Software, and to permit persons to whom the                       
Software is furnished to do so, subject to the following                        
conditions:                                                                     
                                                                                
The above copyright notice and this permission notice shall be                  
included in all copies or substantial portions of the Software.                 
                                                                                
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,                 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES                 
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                        
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT                     
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,                    
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING                    
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR                   
OTHER DEALINGS IN THE SOFTWARE.                                                 
                                                                                
       The author may be contacted at:                                          
           t2ti.com@gmail.com                                                   
                                                                                
@author Albert Eije (alberteije@gmail.com)                    
@version 1.0.0
*******************************************************************************/
using Microsoft.AspNetCore.Mvc;
using System;
using System.Collections.Generic;
using T2TiERPFenix.Models;
using T2TiERPFenix.Repository;

namespace T2TiERPFenix.Controllers
{
    [Route("fin-configuracao-boleto")]
    [Produces("application/json")]
    public class FinConfiguracaoBoletoController : Controller
    {
        private IRepositoryWrapper _repository;

        public FinConfiguracaoBoletoController(IRepositoryWrapper repository)
        {
            _repository = repository;
        }

        [HttpGet]
        public IActionResult ConsultarListaFinConfiguracaoBoleto([FromQuery]string filter)
        {
            try
            {
                IEnumerable<FinConfiguracaoBoleto> lista;
                if (filter == null)
                {
                    lista = _repository.FinConfiguracaoBoleto.ConsultarLista();
                }
                else
                {
                    // define o filtro
                    Filtro filtro = new Filtro(filter);
                    lista = _repository.FinConfiguracaoBoleto.ConsultarListaFiltro(filtro);
                }
                return Ok(lista);
            }
            catch (Exception ex)
            {
                return StatusCode(500, new RetornoJsonErro(500, "Erro no Servidor [Consultar Lista FinConfiguracaoBoleto]", ex));
            }
        }

        [HttpGet("{id}", Name = "ConsultarObjetoFinConfiguracaoBoleto")]
        public IActionResult ConsultarObjetoFinConfiguracaoBoleto(int id)
        {
            try
            {
                var objeto = _repository.FinConfiguracaoBoleto.ConsultarObjeto(id);

                if (objeto == null)
                {
                    return StatusCode(404, new RetornoJsonErro(404, "Registro não localizado [Consultar Objeto FinConfiguracaoBoleto]", null));
                }
                else
                {
                    return Ok(objeto);
                }
            }
            catch (Exception ex)
            {
                return StatusCode(500, new RetornoJsonErro(500, "Erro no Servidor [Consultar Objeto FinConfiguracaoBoleto]", ex));
            }
        }

        [HttpPost]
        public IActionResult InserirFinConfiguracaoBoleto([FromBody]FinConfiguracaoBoleto objJson)
        {
            try
            {
                if (!ModelState.IsValid)
                {
                    return StatusCode(400, new RetornoJsonErro(400, "Objeto inválido [Inserir FinConfiguracaoBoleto]", null));
                }
                _repository.FinConfiguracaoBoleto.Inserir(objJson);

                return CreatedAtRoute("ConsultarObjetoFinConfiguracaoBoleto", new { id = objJson.Id }, objJson);
            }
            catch (Exception ex)
            {
                return StatusCode(500, new RetornoJsonErro(500, "Erro no Servidor [Inserir FinConfiguracaoBoleto]", ex));
            }
        }

        [HttpPut("{id}")]
        public IActionResult AlterarFinConfiguracaoBoleto([FromBody]FinConfiguracaoBoleto objJson, int id)
        {
            try
            {
                if (!ModelState.IsValid)
                {
                    return StatusCode(400, new RetornoJsonErro(400, "Objeto inválido [Alterar FinConfiguracaoBoleto]", null));
                }

                if (objJson.Id != id)
                {
                    return StatusCode(400, new RetornoJsonErro(400, "Objeto inválido [Alterar FinConfiguracaoBoleto] - ID do objeto difere do ID da URL.", null));
                }

                var objBanco = _repository.FinConfiguracaoBoleto.ConsultarObjeto(objJson.Id);

                if (objBanco == null)
                {
                    return StatusCode(400, new RetornoJsonErro(400, "Objeto com ID inválido [Alterar FinConfiguracaoBoleto]", null));
                }

                _repository.FinConfiguracaoBoleto.Alterar(objBanco, objJson);

                return ConsultarObjetoFinConfiguracaoBoleto(id);
            }
            catch (Exception ex)
            {
                return StatusCode(500, new RetornoJsonErro(500, "Erro no Servidor [Alterar FinConfiguracaoBoleto]", ex));
            }
        }

        [HttpDelete("{id}")]
        public IActionResult ExcluirFinConfiguracaoBoleto(int id)
        {
            try
            {
                var objeto = _repository.FinConfiguracaoBoleto.ConsultarObjeto(id);

                _repository.FinConfiguracaoBoleto.Excluir(objeto);

                return Ok();
            }
            catch (Exception ex)
            {
                return StatusCode(500, new RetornoJsonErro(500, "Erro no Servidor [Excluir FinConfiguracaoBoleto]", ex));
            }
        }

    }
}